substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1059 @@
+			// As ticket ids follow a uniform random distribution we expect to
+			// drop more or less half of the segment tickets here.
+			for ticket_id in segment {
+				if ticket_id < upper_bound {
+					sorted_candidates.push(ticket_id);
+				} else {
+					TicketsData::<T>::remove(ticket_id);
+				}
+			}
+
+			if sorted_candidates.len() > max_tickets {
+				// Sort, truncate good tickets, cleanup storage.
+				require_sort = false;
+				sorted_candidates.sort_unstable();
+				sorted_candidates[max_tickets..].iter().for_each(TicketsData::<T>::remove);
+				sorted_candidates.truncate(max_tickets);
+				upper_bound = sorted_candidates[max_tickets - 1];
+			}
+		}
+
+		if require_sort {
+			sorted_candidates.sort_unstable();
+		}
+
+		if metadata.unsorted_tickets_count == 0 {
+			// Sorting is over, write to next epoch map.
+			sorted_candidates.iter().enumerate().for_each(|(i, id)| {
+				TicketsIds::<T>::insert((epoch_tag, i as u32), id);
+			});
+			metadata.tickets_count[epoch_tag as usize] = sorted_candidates.len() as u32;
+		} else {
+			// Keep the partial result for next calls.
+			SortedCandidates::<T>::set(BoundedVec::truncate_from(sorted_candidates));
+		}
+	}
+
+	/// Append a set of tickets to the segments map.
+	pub(crate) fn append_tickets(tickets: BoundedVec<TicketId, MaxTicketsFor<T>>) {
+		debug!(target: LOG_TARGET, "Appending batch with {} tickets", tickets.len());
+		tickets.iter().for_each(|t| debug!(target: LOG_TARGET, "  + {t:032x}"));
+
+		let mut metadata = TicketsMeta::<T>::get();
+		let mut segment_idx = metadata.unsorted_tickets_count / SEGMENT_MAX_SIZE;
+
+		let mut tickets = tickets.as_slice();
+		while !tickets.is_empty() {
+			let rem = metadata.unsorted_tickets_count % SEGMENT_MAX_SIZE;
+			let todo = tickets.len().min((SEGMENT_MAX_SIZE - rem) as usize);
+
+			let mut segment = UnsortedSegments::<T>::get(segment_idx).into_inner();
+			segment.extend_from_slice(&tickets[..todo]);
+			let segment = BoundedVec::truncate_from(segment);
+			UnsortedSegments::<T>::insert(segment_idx, segment);
+
+			metadata.unsorted_tickets_count += todo as u32;
+			segment_idx += 1;
+
+			tickets = &tickets[todo..];
+		}
+
+		TicketsMeta::<T>::set(metadata);
+	}
+
+	/// Remove all tickets related data.
+	///
+	/// May not be efficient as the calling places may repeat some of this operations
+	/// but is a very extraordinary operation (hopefully never happens in production)
+	/// and better safe than sorry.
+	fn reset_tickets_data() {
+		let tickets_metadata = TicketsMeta::<T>::get();
+
+		// Remove even-epoch data.
+		let tickets_count = tickets_metadata.tickets_count[0];
+		for idx in 0..tickets_count {
+			if let Some(id) = TicketsIds::<T>::get((0, idx)) {
+				TicketsData::<T>::remove(id);
+			}
+		}
+
+		// Remove odd-epoch data.
+		let tickets_count = tickets_metadata.tickets_count[1];
+		for idx in 0..tickets_count {
+			if let Some(id) = TicketsIds::<T>::get((1, idx)) {
+				TicketsData::<T>::remove(id);
+			}
+		}
```

Comment:
It could be squashed into 0..=1 loop, code looks the same.

Reply:
64406b61616e5bfad81a02e56610431ea7e18e4a